pkg/v3/tickers/sample.go

```diff
@@ -0,0 +1,150 @@
+package tickers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync/atomic"
+	"time"
+
+	"github.com/smartcontractkit/ocr2keepers/internal/util"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+type ratio interface {
+	// OfInt should return n out of x such that n/x ~ r (ratio)
+	OfInt(int) int
+}
+
+type shuffler[T any] interface {
+	Shuffle([]T) []T
+}
+
+type upkeepsGetter interface {
+	GetActiveUpkeeps(context.Context, ocr2keepers.BlockKey) ([]ocr2keepers.UpkeepPayload, error)
+}
+
+type sampleTicker struct {
+	// provided dependencies
+	observer      observer
+	getter        upkeepsGetter
+	ratio         ratio
+	samplingLimit time.Duration
+	logger        *log.Logger
+
+	// set by constructor
+	blocks   *BlockTicker
+	shuffler shuffler[ocr2keepers.UpkeepPayload]
+
+	// run state
+	running atomic.Bool
+	chClose chan struct{}
+}
+
+func (st *sampleTicker) Start(ctx context.Context) error {
+	if st.running.Load() {
+		return fmt.Errorf("already running")
+	}
+
+	go func() {
+		if err := st.blocks.Start(ctx); err != nil {
+			st.logger.Printf("error starting block ticker: %s", err)
+		}
+	}()
+
+	st.running.Store(true)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+Loop:
+	for {
+		select {
+		case h := <-st.blocks.C:
+			if len(h) == 0 {
+				continue
+			}
+
+			ctx, cancelFn := context.WithTimeout(ctx, st.samplingLimit)
+
+			// do the observation with limited time
+			tick, err := st.getterFn(ctx, h[0])
```

Comment:
n00b: why we get only the first element?

Reply:
Since BlockHistory is an array of BlockKey, we need the latest BlockKey. Ideally we should create a helper function for this on BlockHistory. I'll do that.